### ISSUE TITLE
Train/Val Loss Issues when training GPT-2 from OWT

### ISSUE THREAD

**Issue by JustinKunzi** [karpathy/nanoGPT#303]
I'm currently testing a new system of 4x NVIDIA A4500's and had no problem running the Shakespeare training set and the outcome was as intended. My system is currently training the GPT 124M Model on the OWT dataset and everything seemed perfectly fine until about 20 steps in when our loss value increased until reaching a value of ~7.81 and its been hovering within ~0.5 of that value for around 65 steps now. Checking the Loss graph on the readme this seems highly unusual and could be indicating some form of problem. Any ideas what could be happening? Here is the config changes that were made to efficiently utilize the most VRAM to reduce total training time. 

batch_size = 16
block_size = 1024
gradient_accumulation_steps = 5 * 4

Gradient steps were multiplied by 4 since our node only has 4 gpu's and the batch size was increased to 16 since the original config at 12 was utilizing only 60% of GPU memory and an increase to 16 changed that to around 94% and sped up the iteration speed. This is the ONLY code that was changed before starting training. I've also included the loss graphs from wandb to help visualize the problem.
![W B Chart 6_17_2023, 9_18_27 AM](https://github.com/karpathy/nanoGPT/assets/114789147/6a8d276b-b2a7-459f-890f-10e1662b0710)
![W B Chart 6_17_2023, 9_18_37 AM](https://github.com/karpathy/nanoGPT/assets/114789147/49f9030e-4443-4819-b6cf-3149393bbd51)

If anyone could explain what could possibly be going on that would be much appreciated. Thank you in advance.

**Comment by JustinKunzi** [comment]
Update: Its day 5/11 of training the GPT-2 on OWT our loss values have still not moved much from the original jump. Our training and loss values are still almost identical and I'm still not entirely sure what is going on. I'm certain that its not overfitting as it occurred so early in training as well as our validation and training losses are so close never separating more than ~0.2 away from each other. I've included both updated loss graphs to again help visualize what is going on. 
![W B Chart 6_20_2023, 3_35_39 PM](https://github.com/karpathy/nanoGPT/assets/114789147/cd8ebb72-435e-4f78-802e-423724ce504e)
![W B Chart 6_20_2023, 3_35_31 PM](https://github.com/karpathy/nanoGPT/assets/114789147/48d17373-89be-42b1-8cac-38c575afc414)

**Comment by huoyushequ** [comment]
I encounter the same issue, the train loss suddenly jump from ~3.0 to ~7.0, and then stay around 7.0 for a long time. Did you figure out the issue?

**Comment by JustinKunzi** [comment]
> I encounter the same issue, the train loss suddenly jump from ~3.0 to ~7.0, and then stay around 7.0 for a long time. Did you figure out the issue?

Unfortunately not. The training finished having stayed within 0.5 of 7.0 nearly the entire time. Its odd that you have the exact same loss value problem. Perhaps something with the dataset then? I see a lot of people on here are doing the simple Shakespeare or having other issues. I haven't read of someone having a successful attempt yet at training the full OWT dataset. I'm thinking about training on a new dataset in the near future when I get some free time. Hopefully then the issue doesn't persist.

**Comment by iainmelvin** [comment]
This happened to me too, with one A100 on openwebtext with default parameters. If anyone knows what might be going on I am all ears!

**Comment by jiacheng-ye** [comment]
Same issue for me. I'm using 4*A100 80G, I change the batch 12 -> 24 and gradient_accumulation_steps = 5 * 8 -> 5*4.

**Comment by jiacheng-ye** [comment]
disable the flash attention works for me

**Comment by ziqi-zhang** [comment]
I'm training GPT-2 (124MB) on OpenWebText and I encountered the same problem. Did you figure out it?

**Comment by karan78tp** [comment]
Hi,
I am using around 42GB GPU. dataset = open web text (20GB).
batch size = 16 i am training it for 8th day now.
![Screenshot_2023-08-08-13-06-40-82_40deb401b9ffe8e1df2f1cc5ba480b12](https://github.com/karpathy/nanoGPT/assets/36001854/bb8d9b87-d677-441b-9965-71269c9b56a7)
val Loss at 2.95

**Comment by ziqi-zhang** [comment]
@karan78tp Can you share the hyperparameter setting in the config/train_gpt2.py? Thanks!

**Comment by karan78tp** [comment]
@ziqi-zhang batch size is 20.




# config for training GPT-2 (124M) down to very nice loss of ~2.85 on 1 node of 8X A100 40GB
# launch as the following (e.g. in a screen session) and wait ~5 days:
# $ torchrun --standalone --nproc_per_node=8 train.py config/train_gpt2.py

wandb_log = True
wandb_project = 'owt'
wandb_run_name='gpt2-124M'

# these make the total batch size be ~0.5M
# 12 batch size * 1024 block size * 5 gradaccum * 8 GPUs = 491,520
batch_size = 20
block_size = 1024
gradient_accumulation_steps = 5 * 8

# this makes total number of tokens be 300B
max_iters = 600000
lr_decay_iters = 600000

# eval stuff
eval_interval = 1000
eval_iters = 200
log_interval = 10

# weight decay
weight_decay = 1e-1

**Comment by ziqi-zhang** [comment]
@karan78tp Thanks! I was also wondering how many GPUs do you use? Do you use only one GPU?

**Comment by karan78tp** [comment]
i am using two GPU Quadro RTX 6000 each 24576MiB

**Comment by ziqi-zhang** [comment]
Thanks!

**Comment by Tony-Hou** [comment]
It is gradient_accumulation_steps cause the issuse, keep gradient_accumulation_steps = 8  train is ok!

**Comment by karan78tp** [comment]
converged configurations : batch size= 20 ,gradient steps =  40
original config                     batch size = 12 , gradient steps = 40
earlier configurations which failed to converge.
i) batch size =  16 , gradient steps = 20
ii) batch size =  24  , gradient steps  = 40
i guess may be the ratio of gradient/batch size should be more.

**Comment by TahaBinhuraib** [comment]
@karan78tp this is with 2 gpus right?

**Comment by zzw-zwzhang** [comment]
> Hi, I am using around 42GB GPU. dataset = open web text (20GB). batch size = 16 i am training it for 8th day now. ![Screenshot_2023-08-08-13-06-40-82_40deb401b9ffe8e1df2f1cc5ba480b12](https://private-user-images.githubusercontent.com/36001854/259025684-bb8d9b87-d677-441b-9965-71269c9b56a7.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY2MzMwODIsIm5iZiI6MTcwNjYzMjc4MiwicGF0aCI6Ii8zNjAwMTg1NC8yNTkwMjU2ODQtYmI4ZDliODctZDY3Ny00NDFiLTk5NjUtNzEyNjljOWI1NmE3LmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTMwVDE2Mzk0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQwYzU1MTZmN2E2NzIzN2ExNDQxZjVkYmMyMzM0NjQ0NjllMmEzZjM5MDlkMjBiZmFjMzFhYWQ2NjAwZWFlYjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.5rJojb4rIywZvhJDr8L6RCiKeuIvbYL-dKdNWNIezL0) val Loss at 2.95

Can you share your wandb log?

**Comment by iminfine** [comment]
Disable flash attention helps converge.

**Comment by abs-hello** [comment]
> Update: Its day 5/11 of training the GPT-2 on OWT our loss values have still not moved much from the original jump. Our training and loss values are still almost identical and I'm still not entirely sure what is going on. I'm certain that its not overfitting as it occurred so early in training as well as our validation and training losses are so close never separating more than ~0.2 away from each other. I've included both updated loss graphs to again help visualize what is going on.
> 
> ![W B Chart 6_20_2023, 3_41_09 PM](https://github.com/karpathy/nanoGPT/assets/114789147/33158c76-d3be-4fef-8bed-0cf39b1d9fe2) ![W B Chart 6_20_2023, 3_35_31 PM](https://github.com/karpathy/nanoGPT/assets/114789147/48d17373-89be-42b1-8cac-38c575afc414)

Hi，did you change any original parameters?
